img2dataset/downloader.py

```diff
@@ -34,6 +36,69 @@ def is_disallowed(headers, user_agent_token, disallowed_header_directives):
     return False
 
 
+class RobotFileParser(urllib.robotparser.RobotFileParser):
+    """Extends RobotFileParser to handle Request urls, and to parse responses more in line with RFC9309"""
+
+    def __init__(self, url=""):
+        super().__init__()
+        self.set_url(url)
+
+    def set_url(self, url):
+        """Sets the URL referring to a robots.txt file, from either a url string or a Request object"""
+        self.url = url
+        if isinstance(url, urllib.request.Request):
+            self.host = url.host
+            self.path = url.selector
+        else:
+            self.host, self.path = urllib.parse.urlparse(url)[1:3]
+
+    def read(self):
+        """Reads the robots.txt URL and feeds it to the parser."""
+        try:
+            with urllib.request.urlopen(self.url, timeout=10) as f:
+                if f.info().get_content_type() != "text/plain":
+                    # server did not respond with a parsable robots.txt file, handle the same as unavailable status
+                    self.allow_all = True
+                else:
+                    raw = f.read()
+                    self.parse(raw.decode("utf-8").splitlines())
+        except urllib.error.HTTPError as err:
+            if err.code in (401, 403) or 500 <= err.code <= 599:
+                # server is unreachable, deny all
+                self.disallow_all = True
+            elif 400 <= err.code <= 499:
+                # robots.txt is unavailable, allow all (even though future requests are likely to fail)
+                self.allow_all = True
+        except Exception:  # pylint: disable=broad-except
+            # treat other errors as meaning the server is unreachable (timeout, ssl error, dns failure)
+            self.disallow_all = True
+
+
+thread_context = local()
+
+
+def init_thread_context():
+    thread_context.robots_txt_cache = {}
+
+
+def can_fetch(url, user_agent):
+    """Returns True if the given user agent is allowed to fetch this url based on the hosts robots.txt"""
+    robots_txt_cache = thread_context.robots_txt_cache
+
+    robots_url = urllib.parse.urljoin(url, "/robots.txt")
```

Comment:
this does not seem correct
   The rules MUST be accessible in a file named "/robots.txt" (all
   lowercase) in the top-level path of the service. 

https://datatracker.ietf.org/doc/rfc9309/ 2.3

url is not the top-level path of the service
it would be needed to keep only the domain name

Reply:
https://tio.run/##TcgxDoMwDAXQnVOgTCBV9tCN2wRqkaAk/nJciZ4@7djxPXw8aXuOkSvUfH5bKXknROsyTbDcfPk/@uHS3JaQ3NE3ZrljRRE6tHLXKvwSASN6ogtneMyBTXf1Tn57WNcxvg

"/robots.txt" is a root-relative URL. urljoin with a root relative URL as the second argument uses the host, scheme, and port from the first argument to make a complete URL.